service/worker/archiver/history_index_blob.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package archiver
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/dgryski/go-farm"
+	"github.com/uber/cadence/common/blobstore/blob"
+)
+
+var (
+	errNoKnownVersions = errors.New("history index blob contains no versions")
+)
+
+// NewHistoryIndexBlobKey returns a key for history index blob
+func NewHistoryIndexBlobKey(domainID, workflowID, runID string) (blob.Key, error) {
+	if len(domainID) == 0 || len(workflowID) == 0 || len(runID) == 0 {
+		return nil, errInvalidKeyInput
+	}
+	domainIDHash := fmt.Sprintf("%v", farm.Fingerprint64([]byte(domainID)))
+	workflowIDHash := fmt.Sprintf("%v", farm.Fingerprint64([]byte(workflowID)))
+	runIDHash := fmt.Sprintf("%v", farm.Fingerprint64([]byte(runID)))
+	combinedHash := strings.Join([]string{domainIDHash, workflowIDHash, runIDHash}, "")
+	return blob.NewKey("index", combinedHash)
+}
+
+// GetHighestVersion returns the highest version from index blob tags
+func GetHighestVersion(tags map[string]string) (*int64, error) {
```

Comment:
I return a pointer so that I can return nil if there is an error. Rather than returning some token value like -1.

Reply:
But there's already an error output and we don't need to perform nil pointer check to determine if there's an error or not.

Reply:
+1 preferable to return an `int64`, so that the caller does not have to run a nil check.. Less prone to error..

Reply:
The reason I return a pointer is -1 is not really a token value. In single DC case failover version can be negative. If I return -1 as a token value and the user ignores the error using this token value it can mess things up. I prefer to only return a value if its valid.

As for the nil check, clients of this API will not do the nil check. The pattern I have seen in other places is you check if error is not nil and if that check passes you just assume the result is not nil.